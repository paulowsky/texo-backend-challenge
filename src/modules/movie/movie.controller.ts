import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  Param,
  Patch,
  Post
} from '@nestjs/common'

import { CreateMovieInput, UpdateMovieInput } from './dto/movie.input'
import { Movie } from './entities/movie.entity'
import { MovieService } from './movie.service'

@Controller('movie')
export class MovieController {
  constructor(private movieService: MovieService) {}

  @HttpCode(200)
  @Get('')
  async getAll(): Promise<Movie[]> {
    return this.movieService.getAll()
  }

  @HttpCode(200)
  @Get(':title')
  async getOne(@Param('title') title: string): Promise<Movie> {
    return this.movieService.getOne(title)
  }

  @HttpCode(201)
  @Post('')
  async create(@Body() data: CreateMovieInput): Promise<Movie> {
    return this.movieService.create(data)
  }

  @HttpCode(200)
  @Patch('')
  async update(@Body() data: UpdateMovieInput): Promise<Movie> {
    return this.movieService.update(data)
  }

  @HttpCode(200)
  @Delete(':title')
  async delete(@Param('title') title: string): Promise<void> {
    return this.movieService.delete(title)
  }
}
