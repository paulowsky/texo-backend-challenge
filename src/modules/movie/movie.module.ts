import { Module } from '@nestjs/common'

import { DatabaseModule } from '@db/db.module'

import { MovieController } from './movie.controller'
import { MovieService } from './movie.service'

@Module({
  imports: [DatabaseModule],
  controllers: [MovieController],
  providers: [MovieService],
  exports: [MovieService]
})
export class MovieModule {}
