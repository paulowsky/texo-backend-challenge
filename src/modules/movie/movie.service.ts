import { Injectable, NotFoundException } from '@nestjs/common'

import { DatabaseMovie, DatabaseService } from '@db/db.service'

import { CreateMovieInput, UpdateMovieInput } from './dto/movie.input'

@Injectable()
export class MovieService {
  constructor(private readonly db: DatabaseService) {}

  async getAll(filter?: Partial<DatabaseMovie>) {
    const result = await this.db.findAllMovies(filter)

    const movies = result.map(movie => ({
      ...movie,
      year: parseInt(movie.year),
      studios: movie.studios.split(', ').map(str => str.trim()),
      producers: movie.producers
        .replace(' and ', ', ')
        .split(', ')
        .map(str => str.trim()),
      winner: movie.winner === 'yes' ? true : false
    }))

    return movies
  }

  async getOne(title: string) {
    const result = await this.getAll({ title })

    if (!result.length) throw new NotFoundException('Movie not found')

    return result[0]
  }

  async create(data: CreateMovieInput) {
    const normalizedData = {
      title: data.title,
      winner: data.winner ? 'yes' : '',
      year: data.year.toString(),
      studios: data.studios.join(', '),
      producers: data.producers.join(', ')
    }

    await this.db.createMovie(normalizedData)

    return await this.getOne(normalizedData.title)
  }

  async update(data: UpdateMovieInput) {
    await this.getOne(data.originalTitle)

    const normalizedData = {
      title: data.title,
      winner: data.winner ? 'yes' : '',
      year: data.year.toString(),
      studios: data.studios.join(', '),
      producers: data.producers.join(', ')
    }

    await this.db.updateMovie(data.originalTitle, normalizedData)

    return await this.getOne(
      normalizedData.title ? normalizedData.title : data.originalTitle
    )
  }

  async delete(title: string) {
    await this.getOne(title)

    await this.db.deleteMovie(title)
  }
}
