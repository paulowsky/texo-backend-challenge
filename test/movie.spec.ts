import { INestApplication } from '@nestjs/common'
import { Test } from '@nestjs/testing'
import * as request from 'supertest'

import { MovieController } from '@modules/movie/movie.controller'
import { MovieService } from '@modules/movie/movie.service'

import { mockMovieService, testMovie } from './fixtures/mocks'

describe('Movie Module', () => {
  let app: INestApplication

  beforeAll(async () => {
    const moduleFixture = await Test.createTestingModule({
      controllers: [MovieController],
      providers: [
        {
          provide: MovieService,
          useValue: mockMovieService
        }
      ]
    }).compile()

    app = moduleFixture.createNestApplication()

    await app.init()
  })

  afterAll(async () => {
    await app.close()
  })

  it('(GET) /movie', async () => {
    const response = await request(app.getHttpServer()).get('/movie')

    expect(response.statusCode).toEqual(200)
    expect(response.body).toMatchSnapshot()
  })

  it('(GET) /movie/:title', async () => {
    const response = await request(app.getHttpServer()).get(
      `/movie/${testMovie.title}`
    )

    expect(response.statusCode).toEqual(200)
    expect(response.body).toMatchSnapshot()
  })

  it('(POST) /movie', async () => {
    const response = await request(app.getHttpServer())
      .post('/movie')
      .send(testMovie)

    expect(response.statusCode).toEqual(201)
    expect(response.body).toMatchSnapshot()
  })

  it('(PATCH) /movie', async () => {
    const response = await request(app.getHttpServer())
      .patch('/movie')
      .send({ ...testMovie, title: 'edit', originalTitle: testMovie.title })

    expect(response.statusCode).toEqual(200)
    expect(response.body).toMatchSnapshot()
  })

  it('(DELETE) /movie/:title', async () => {
    const response = await request(app.getHttpServer()).delete(
      `/movie/${testMovie.title}`
    )

    expect(response.statusCode).toEqual(200)
    expect(response.body).toMatchSnapshot()
  })
})
