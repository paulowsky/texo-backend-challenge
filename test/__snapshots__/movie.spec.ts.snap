// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Module (DELETE) /movie/:title 1`] = `{}`;

exports[`Movie Module (GET) /movie 1`] = `
[
  {
    "producers": [
      "test",
    ],
    "studios": [
      "test",
    ],
    "title": "test",
    "winner": false,
    "year": 1999,
  },
]
`;

exports[`Movie Module (GET) /movie/:title 1`] = `
{
  "producers": [
    "test",
  ],
  "studios": [
    "test",
  ],
  "title": "test",
  "winner": false,
  "year": 1999,
}
`;

exports[`Movie Module (PATCH) /movie 1`] = `
{
  "producers": [
    "test",
  ],
  "studios": [
    "test",
  ],
  "title": "edit",
  "winner": false,
  "year": 1999,
}
`;

exports[`Movie Module (POST) /movie 1`] = `
{
  "producers": [
    "test",
  ],
  "studios": [
    "test",
  ],
  "title": "test",
  "winner": false,
  "year": 1999,
}
`;
